model/vm_pool.rb

```diff
@@ -13,13 +14,9 @@ class VmPool < Sequel::Model
 
   def pick_vm
     DB.transaction do
-      # first lock the whole pool in the join table so that no other thread can
-      # pick a vm from this pool
-      vms_dataset.for_update.all
-      pick_vm_id_q = vms_dataset.left_join(:github_runner, vm_id: :id)
-        .where(Sequel[:github_runner][:vm_id] => nil, Sequel[:vm][:display_state] => "running")
-        .select(Sequel[:vm][:id])
-      Vm.where(id: pick_vm_id_q).first
+      # We lock the idle vms in the pool so that no other thread can pick the same
+      locked_vms = idle_vms_dataset.for_update.order(:created_at).all
```

Comment:
I'm thinking if we should perform the `name` and `tainted_at` update here and `rescue` in case `.update().zero?` and recursively call pick_vm. This would also mean we will need to add another line to check if pool is empty and return `nil` in that case.

Reply:
Thanks for recommendation. I refactored it. We avoid locking with the new version 🎉 

I tested the new approach with the following snippet and the first pry raised an exception.

```
# pry 1st
p = VmPool.first
DB.transaction do
    vm = p.idle_vms_dataset.order(:created_at).first
    puts "picked #{vm.name}"
    sleep 20
    fail "concurrent pick" if p.idle_vms_dataset.where(id: vm.id).update(tainted_at: Time.now).zero?
    puts "selected #{vm.name}"
end


# pry 2nd
p = VmPool.first
DB.transaction do
    vm = p.idle_vms_dataset.order(:created_at).first
    puts "picked #{vm.name}"
    fail "concurrent pick" if p.idle_vms_dataset.where(id: vm.id).update(tainted_at: Time.now).zero?
    puts "selected #{vm.name}"
end
```